File: 1203-print-in-order/1203-print-in-order.cpp
class Foo {
public:
    mutex mtx;
    condition_variable cv;
    int turn;

    Foo() {
        turn = 1;
    }

    void first(function<void()> printFirst) {
        printFirst();
        turn = 2;
        cv.notify_all();
    }

    void second(function<void()> printSecond) {
        unique_lock<mutex> lock(mtx);
        while(turn != 2) {
            cv.wait(lock);
        }
        printSecond();
        turn = 3;
        cv.notify_all();
    }

    void third(function<void()> printThird) {
        unique_lock<mutex> lock(mtx);
        while(turn != 3) {
            cv.wait(lock);
        }
        printThird();
    }
};

File: 1203-print-in-order/1203-print-in-order.py
class Foo:
    def __init__(self):
        self.turn = 1
        self.cv = threading.Condition()

    def first(self, printFirst: Callable[[], None]) -> None:
        with self.cv:  # Acquire the lock before notifying
            printFirst()
            self.turn = 2
            self.cv.notify_all()  # Notify waiting threads

    def second(self, printSecond: Callable[[], None]) -> None:
        with self.cv:  # Acquire the lock before waiting
            while self.turn != 2:  # Use `while` to recheck condition after wakeup
                self.cv.wait()  # Wait for notification
            printSecond()
            self.turn = 3
            self.cv.notify_all()  # Notify waiting threads

    def third(self, printThird: Callable[[], None]) -> None:
        with self.cv:  # Acquire the lock before waiting
            while self.turn != 3:  # Use `while` to recheck condition after wakeup
                self.cv.wait()  # Wait for notification
            printThird()